### PR TITLE
fix profile upgrade watcher for containers

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

Add a watcher for containers charm profiles charms, refactor modelMachinesProfileWatcher to work with different machine types.

Existing tests work, new tests forthcoming.

## QA steps

1. `export JUJU_DEV_FEATURE_FLAGS=lxd-profile`
2. `juju bootstrap localhost --config logging-config='<root>=DEBUG;unit=DEBUG'`
3. `juju deploy ./testcharms/charm-repo/quantal/lxd-profile lxd-profile-local`
7. `juju add-unit lxd-profile-local --to lxd`
Once deploy is complete, upgrade the charm
5. `juju upgrade-charm lxd-profile-local --path ./testcharms/charm-repo/quantal/lxd-profile`
6. run `'juju debug-log -m controller --include-module juju.provisioner'` watching for debug msg about "processProfileChanges" for the machine
6. run `'juju debug-log --include-module juju.provisioner'` watching for debug msg about "processProfileChanges" for the container

### CODE REVIEW

File: apiserver/facades/client/client/status.go [hunk 1]
```diff
@@ -915,7 +915,7 @@ func (c *statusContext) makeMachineStatus(machine *state.Machine, appStatusInfo
 			}
 		}
 	} else {
-		logger.Debugf("error fetching lxd profiles: %q", err)
+		logger.Debugf("error fetching lxd profiles for %s: %q", machine.String(), err.Error())
```
Comment (hmlanigan) [inline apiserver/facades/client/client/status.go]: drive by: the error wasn't printing out.

File: state/watcher.go [hunk 2]
```diff
@@ -823,27 +858,26 @@ func newModelMachinesProfileWatcher(backend modelBackend) StringsWatcher {
...
 
-	iter := newMachines.Find(nil).Iter()
+	iter := newMachines.Find(w.members).Iter()
 	for iter.Next(&doc) {
+		// If no members criteria is specified, use the filter
+		// to reject any unsuitable initial elements.
+		if w.members == nil && w.filter != nil && !w.filter(doc.Id) {
+			continue
+		}
 		w.known[doc.Id] = doc.UpgradeCharmProfileCharmURL
```
Comment (SimonRichardson) [inline state/watcher.go]: As we talked about earlier, we'll make this more generic in future commits.

File: worker/provisioner/provisioner.go [hunk 3]
```diff
@@ -405,5 +405,13 @@ func (p *containerProvisioner) getRetryWatcher() (watcher.NotifyWatcher, error)
 }
 
 func (p *containerProvisioner) getProfileWatcher() (watcher.StringsWatcher, error) {
-	return p.st.WatchModelMachinesCharmProfiles()
+	if p.containerType != instance.LXD {
+		// TODO (hml) lxd-profile 17-oct-2018
+		// what is the correct way to handle this, only start for LXD containers
```
Comment (SimonRichardson) [inline worker/provisioner/provisioner.go]: I wonder if we can return `nil, nil` or a dumb watcher, that never watches anything, so effectively it becomes a no-op watcher?

**Comment by hmlanigan** [comment]
$$merg$$

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**